File: Python/프로그래머스/level_2/거리두기확인하기.py
#  https://programmers.co.kr/questions/20609

# 1. 사람이 앉은 자리 + 사방에 -1을 해줌
# 2. 칸막이가 있는 곳에는 +10(3 이상의 숫자 아무거나)를 해줌
# 3. 전체를 돌면서 -2 이하의 수가 있으면 거리두기를 안지킨거


def check(matrix):
    for i in range(5):
            for j in range(5):
                if matrix[i][j] <= -2:
                    return 0
                
    return 1

def solution(places):
    answer = []
    
    for place in places:
        matrix = [[0 for _ in range(5)] for _ in range(5)]
        
        ps = []
        xs = []
        
        for i in range(5):
            for j in range(5):
                if place[i][j] == "P":
                    ps.append((i,j))
                elif place[i][j] == "X":
                    xs.append((i,j))
        
        # # 북 -> 동 -> 남 -> 서
        # dr = [-1, -1, 0, 1, 1, 1, 0, -1]
        # dc = [0, 1, 1, 1, 0, -1, -1, -1]
        
        # 앉은자리 + 사방
        dr = [0, 0, 1, 0, -1]
        dc = [0, 1, 0, -1, 0]
        
                    
        for p in ps:
            px, py = p
            for i in range(5):
                nr = px + dr[i]
                nc = py + dc[i]
                
                if not(0<=nr<5 and 0<=nc<5):
                    continue
                    
                
                matrix[nr][nc] -= 1
            
        for x in xs:
            xx, xy = x
            matrix[xx][xy] += 10
        
        answer.append(check(matrix))
        
        
        
    return answer

File: Kotlin/프로그래머스/Level 2/거리두기확인하기.kt
// P: 응시자가 앉음, O: 빈 테이블, X: 파티션
// -> 응시자가 앉은곳 + 4방 -> -1, 파티션 있는 칸 -> +100
// 해서 나온 배열의 음수인 곳이 있나 확인
// 사람이 앉은 위치가 아닌 파티션이 있어야 하는 위치에 파티션이 있나 없나 확인하는것!
class Solution {
    fun solution(places: Array<Array<String>>): IntArray {
        var answer: IntArray = intArrayOf()
        
        for(place in places) {
            var matrix = ArrayList<CharArray>()
            var checkMatrix = ArrayList<IntArray>() // 숫자 체크를 위한 배열
            var drPerson = arrayOf<Int>(-1, 0, 1, 0, 0) // 4방 + 자신
            var dcPerson = arrayOf<Int>(0, 1, 0, -1, 0)
            // var drPartition = arrayOf<Int>(-1, 1, 1, -1) // 대각선 4방
            // var dcPartition = arrayOf<Int>(1, 1, -1, -1)
            
            for(str in place){
                matrix.add(str.toCharArray())
            }
            
            
            for(i in 0 until 5){ // 숫자 체크 배열 초기화
                var newArr = IntArray(5)
                checkMatrix.add(newArr)
            }

            for(i in 0 until 5){ // 사람 앉거나 파티션이 있으면 연산
                for(j in 0 until 5){
                    if(matrix[i][j] == 'P'){ // 사람이 앉았으면
                        for(k in 0 until 5){ // 4방 + 자신
                            var nr = i + drPerson[k]
                            var nc = j + dcPerson[k]
                            
                            if(!(nr in 0 until 5 && nc in 0 until 5)){ // 범위 제한
                                continue
                            }
                            
                            checkMatrix[nr][nc] -= 1 // -1 을 해줌
                        }
                    }
                    else if(matrix[i][j] == 'X'){ // 파티션이 있으면
                        checkMatrix[i][j] += 100 // 100 을 해줌
                    }
                }
            }
            // checkMatrix.forEach{
            //     println(it.joinToString())
            // }
            // println("-".repeat(20))
            
            // 이제 matrix를 돌면서 사람 앉은곳에 음수가 있나 확인
            var flag = true
            base@ for(i in 0 until 5){
                for (j in 0 until 5){
                        if(checkMatrix[i][j] <= -2){
                            // println("${i}, ${j}")
                            answer += 0
                            flag = false
                            break@base
                        }
                    }
            }
            
            if (flag) {
                answer += 1
            }
    }
    return answer
    }
}